src/frontend/ocamlmerlin/new/new_merlin.ml

```diff
@@ -147,17 +147,23 @@ let run = function
         prerr_endline ("Exception: " ^ Printexc.to_string exn);
         1
 
+let restore_cwd old f args =
+  match f args with
+  | exception e -> Sys.chdir old; raise e
+  | res -> Sys.chdir old; res
+
 let run ~new_env wd args =
   begin match new_env with
   | Some env ->
     Os_ipc.merlin_set_environ env;
     Unix.putenv "__MERLIN_MASTER_PID" (string_of_int (Unix.getpid ()))
   | None -> () end;
-  let wd_msg = match wd with
-    | None -> "No working directory specified"
+  let wd_msg, run = match wd with
+    | None -> "No working directory specified", run
     | Some wd ->
-      try Sys.chdir wd; Printf.sprintf "changed directory to %S" wd
-      with _ -> Printf.sprintf "cannot change working directory to %S" wd
+      let oldwd = Sys.getcwd () in
+      try Sys.chdir wd; Printf.sprintf "changed directory to %S" wd, restore_cwd oldwd run
```

Comment:
So, do I understand well that without these change we change the server cwd before running a command without restoring it after that. Which means that the first cd to `/` is useless ?

Could you add a test illustrating this change's effect ?

Reply:
> So, do I understand well that without these changes we update the server cwd before running a command without restoring it after that.

Yes.

> Which means that the first cd to `/` is useless ?

No, it is still necessary: otherwise after handling a request the server is left on the directory where it was first launched, and this can cause problems.

> Could you add a test illustrating this change's effect ?

I'm not sure how I would go about it, since we are talking about correctly (re)setting the "current directory" of the server process, but there should be no observable difference for the user. But if you have any suggestions, I'm happy to give it a try.

Reply:
I pushed a test that shows that the current working directory was not being restored correctly before. I didn't try to exhibit a server misbehaviour because this was harder, it depends on the operating system, etc.

Reply:
Great, thank you !